Sources/Datadog/Logging/RemoteLogger.swift

```diff
@@ -0,0 +1,158 @@
+    /// Integration with RUM. It is used to send RUM errors for logs of high severity (`.error` and above).
+    /// Can be `nil` if RUM feature is disabled.
+    internal let rumErrorsIntegration: LoggingWithRUMErrorsIntegration?
+    /// Integration with Tracing. It is used to correlate Logs with Spans by injecting `Span` context to `LogEvent`.
+    /// Can be `nil` if the integration is disabled for this logger or if Tracing feature is disabled.
+    internal let activeSpanIntegration: LoggingWithActiveSpanIntegration?
+
+    init(
+        core: DatadogCoreProtocol,
+        configuration: Configuration,
+        dateProvider: DateProvider,
+        rumContextIntegration: LoggingWithRUMContextIntegration?,
+        rumErrorsIntegration: LoggingWithRUMErrorsIntegration?,
+        activeSpanIntegration: LoggingWithActiveSpanIntegration?
+    ) {
+        self.core = core
+        self.configuration = configuration
+        self.dateProvider = dateProvider
+        self.queue = DispatchQueue(
+            label: "com.datadoghq.logger-\(configuration.loggerName)",
+            target: .global(qos: .userInteractive)
+        )
+        self.builder = LogEventBuilder(
+            service: configuration.service,
+            loggerName: configuration.loggerName,
+            sendNetworkInfo: configuration.sendNetworkInfo,
+            eventMapper: configuration.eventMapper
+        )
+        self.rumContextIntegration = rumContextIntegration
+        self.rumErrorsIntegration = rumErrorsIntegration
+        self.activeSpanIntegration = activeSpanIntegration
+    }
+
+    // MARK: - Attributes
+
+    func addAttribute(forKey key: AttributeKey, value: AttributeValue) {
+        queue.async { self.unsafeAttributes[key] = value }
+    }
+
+    func removeAttribute(forKey key: AttributeKey) {
+        queue.async { self.unsafeAttributes.removeValue(forKey: key) }
+    }
+
+    // MARK: - Tags
+
+    func addTag(withKey key: String, value: String) {
+        queue.async { self.unsafeTags.insert("\(key):\(value)") }
+    }
+
+    func removeTag(withKey key: String) {
+        queue.async { self.unsafeTags = self.unsafeTags.filter { !$0.hasPrefix("\(key):") } }
+    }
+
+    func add(tag: String) {
+        queue.async { self.unsafeTags.insert(tag) }
+    }
+
+    func remove(tag: String) {
+        queue.async { self.unsafeTags.remove(tag) }
+    }
+
+    // MARK: - Logging
+
+    func log(level: LogLevel, message: String, error: Error?, attributes: [String: Encodable]?) {
+        guard level.rawValue >= configuration.threshold.rawValue else {
+            return
+        }
+
+        // on user thread:
+        let date = dateProvider.now
+        let threadName = getCurrentThreadName()
+
+        // SDK context must be requested on the user thread to ensure that it provides values
+        // that are up-to-date for the caller.
+        self.core.v1.scope(for: LoggingFeature.self)?.eventWriteContext { context, writer in
```

Comment:
@maxep Given what we discussed (that `core.scope(for:).eventWriteContext { }` will be always called on the user thread), I now wonder if it would make sense to include both `date` and `threadName` in `DatadogContext` (not now, but later). This way, the `context` might provide:
```swift
context.threadName    // the name of the user thread
context.date          // current device time
context.serverOffset  // correction to server time
context.serverDate    // device time corrected by server time
```

Reply:
Yes, we could do that before jumping on the context thread [here](https://github.com/DataDog/dd-sdk-ios/pull/916/files#diff-8ea90fa9b289cf85fe53811d1d1a3f08d4a33cc7879ac74ff3ea71b9257a5a8dR73-R79) 👍